surface_reconstruction/nksr/nksr_fvdb/sparse_feature_hierarchy.py

```diff
@@ -0,0 +1,333 @@
+# Copyright Contributors to the OpenVDB Project
+# SPDX-License-Identifier: Apache-2.0
+
+from dataclasses import dataclass, field
+from enum import Enum
+
+import torch
+from fvdb import GridBatch, JaggedTensor
+
+from .coord_xform import CoordXform, voxel_center_aligned_coarsening_xform
+
+
+class VoxelStatus(Enum):
+    """
+    Status of a voxel in the structure pruning stage.
+    """
+
+    # The voxel should not exist (pruned)
+    VS_NON_EXIST = 0
+
+    # The voxel exists, but would not have children
+    VS_EXIST_STOP = 1
+
+    # The voxel exists, and would have children
+    VS_EXIST_CONTINUE = 2
+
+
+def evaluate_voxel_status(target_grid: GridBatch, coarse_grid: GridBatch, fine_grid: GridBatch | None) -> torch.Tensor:
+    """Compute per-voxel status labels for structure prediction training.
+
+    Classifies each voxel in target_grid by comparing against coarse/fine hierarchy levels:
+      - VS_NON_EXIST: voxel not in coarse_grid (should be pruned)
+      - VS_EXIST_STOP: voxel in coarse_grid but not in fine_grid (leaf node)
+      - VS_EXIST_CONTINUE: voxel in both grids (has children, should subdivide)
+
+    When fine_grid is None (i.e., evaluating the finest level), all existing voxels are
+    labeled EXIST_STOP since there's no finer level to continue into. When fine_grid is
+    provided, voxels that would have children in the finer level are labeled EXIST_CONTINUE.
+
+    Args:
+        target_grid: Grid defining the voxels to classify.
+        coarse_grid: Reference grid at this hierarchy level.
+        fine_grid: Reference grid one level finer (2x resolution), or None for finest level.
+
+    Returns:
+        uint8 tensor of shape (target_grid.total_voxels,) with VoxelStatus values.
+    """
+    device = target_grid.device
+    if coarse_grid.device != device:
+        raise ValueError(f"Device not match {device} vs {coarse_grid.device}.")
+
+    if fine_grid is not None and fine_grid.device != device:
+        raise ValueError(f"Device not match {device} vs {fine_grid.device}.")
+
+    # Start with a flat tensor of all voxels in the target grid, initialized to NON_EXIST.
+    status = torch.full((target_grid.total_voxels,), VoxelStatus.VS_NON_EXIST.value, dtype=torch.uint8, device=device)
+
+    # Set the intersection of the coarse grid and the target grid to EXIST_STOP.
+    coarse_exist_idx = target_grid.ijk_to_index(coarse_grid.ijk, cumulative=True)
```

Comment:
You can (and should) do this with `inject_to`. Something like (didn't check the docs and wrote this off the top of my head so double check).

```python
status = target_grid.inject_to(coarse_grid, torch.full([coarse_grid.total_voxels,], VoxelStatus.VS_EXIST_STOP.value, dtype=torch.uint8, device=device), default_value=VoxelStatus.VS_NON_EXIST.value)
```

As an aside, it's worth checking if inject will work with expand so you don't need the allocation. If not file an issue.  Specifically, you should check if you can pass the source sidecar as `torch.tensor([VoxelStatus.VS_EXIST_STOP.value]).expand(coarse_grid.total_voxels)`